src/Polysemy/Fresh.hs

```diff
@@ -0,0 +1,141 @@
+--      as well as interpreters for 'Polysemy.Async.Async'.
+--
+-- Prefer 'freshToIO' whenever possible.
+-- If you can't use 'runFreshUnsafe' safely, nor use 'freshToIO', consider
+-- 'runFreshUnsafePerformIO'.
+runFreshUnsafe :: Sem (Fresh UniqueRef ': r) a
+               -> Sem r a
+runFreshUnsafe =
+    (fmap snd .)
+  $ (runState (UniqueRef 0) .)
+  $ reinterpret
+  $ \Fresh -> do
+    UniqueRef s <- get
+    put $! UniqueRef (s + 1)
+    return (UniqueRef s)
+{-# INLINE runFreshUnsafe #-}
+
+
+-----------------------------------------------------------------------------
+-- | A variant of 'runFreshUnsafe' where any 'Enum' may be used as the type
+-- of unique objects.
+runFreshEnumUnsafe :: forall n a r
+                    . Enum n
+                   => Sem (Fresh n ': r) a
+                   -> Sem r a
+runFreshEnumUnsafe =
+    (fmap snd .)
+  $ (runState @n (toEnum 0) .)
+  $ reinterpret
+  $ \Fresh -> do
+    s <- get
+    put $! succ s
+    return s
+{-# INLINE runFreshEnumUnsafe #-}
+
+-----------------------------------------------------------------------------
+-- | Runs a 'Fresh' effect through generating 'Unique's using
+-- 'unsafePerformIO'.
+--
+-- Ironically, despite the fact that this uses 'unsafePerformIO', and
+-- 'runFreshUnsafe' uses no unsafe operations whatsoever, this is still
+-- typically safer to use than 'runFreshUnsafe', although 'runFreshUnsafe'
+-- is perhaps more efficient.
+--
+-- The worst thing that this particular use of 'unsafePerformIO' could result
+-- in is the loss of referential transparency, as rerunning an interpreter stack
+-- using 'runFreshUnsafePerformIO' will create different 'Unique's. This should
+-- never matter.
+--
+-- This could be potentially be less efficient than 'runFreshUnsafe'.
+--
+-- If you ever observe that multiple invocations of 'fresh' produce the same
+-- 'Unique' under 'runFreshUnsafePerformIO', then open an issue over at the
+-- GitHub repository.
+runFreshUnsafePerformIO :: Sem (Fresh Unique ': r) a
+                        -> Sem r a
+runFreshUnsafePerformIO = usingSem $ \u ->
+  case decomp u of
+    Right (Weaving Fresh s _ ex _) -> do
+      let !uniq = unsafePerformIO (newUnique' u)
+          {-# NOINLINE uniq #-}
+      pure $ ex (uniq <$ s)
+    Left g -> liftSem (hoist runFreshUnsafePerformIO g)
+-- KingoftheHomeless: I've tried very hard to prevent optimizations from
+-- sharing the call to 'unsafePerformIO'.
+-- The inlining of 'interpret' is so that I can give
+-- 'u' to 'newUnique'', and thus prevent 'uniq' from floating outside the
+-- lambda. This interpreter might even be safe to INLINE, but I'm not taking
+-- any chances.
+{-# NOINLINE runFreshUnsafePerformIO #-}
+
+newUnique' :: Union (Fresh Unique ': r) (Sem (Fresh Unique ': r)) a -> IO Unique
+newUnique' (Union _ _) = newUnique
+{-# NOINLINE newUnique' #-}
+
+newtype UniqueRef = UniqueRef Integer
+  deriving (Eq, Ord, Hashable)
```

Comment:
This stings a lil bit. I feel `Hashable` is very important instance to have, so unfortunately, this incurs an extra dependency.

Reply:
Let's _just_ provide the `Unique` and `Enum` instances, and then not need the dependency?

Reply:
I don't understand. Do you mean removing `runFreshUnsafe`?

Reply:
@isovector Can you clarify what you mean?

Reply:
We don't need this `UniqueRef` thing, right? And if we dont have it then we don't need a dependency on hashable :)

Reply:
`UniqueRef` is meant as a user-facing symbol, so users don't have to cook up their own, and don't have to deal with ambiguity error messages when using the interpreter.

It's nice to have, but you're right in that we don't strictly need it. Personally, I feel that `hashable` is such a small and universal dependency that the cost of using it here is next to nothing. It's also already a transitive dependency, so we don't exactly force users to download more packages. In the end though, it's your decision.

Reply:
Ok, then I'll remove it.